core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -55,82 +61,158 @@
 
    @Inject protected ComponentRef<AdvancedCache<K, V>> cacheRef;
    @Inject protected ComponentRef<CommandsFactory> cf;
    @Inject protected RpcManager rpcManager;
    @Inject protected DistributionManager distributionManager;
 
    private AdvancedCache<K, V> cache;
+   private Address localAddress;
+   private long timeout;
+   private String cacheName;
 
    @Override
    public void start() {
       super.start();
       // Data container entries are retrieved directly, so we don't need to worry about an encodings
       this.cache = AbstractDelegatingCache.unwrapCache(cacheRef.wired()).getAdvancedCache();
+      this.cacheName = cache.getName();
+      this.localAddress = cache.getCacheManager().getAddress();
+      this.timeout = cache.getCacheConfiguration().clustering().remoteTimeout();
    }
 
    @Override
    public void processExpiration() {
-      long start = 0;
+
       if (!Thread.currentThread().isInterrupted()) {
-         try {
-            if (trace) {
-               log.trace("Purging data container of expired entries");
-               start = timeService.time();
-            }
-            // We limit it so there is only so many async expiration removals done at the same time
-            List<CompletableFuture> futures = new ArrayList<>(MAX_ASYNC_EXPIRATIONS);
-            long currentTimeMillis = timeService.wallClockTime();
-            dataContainer.running().forEachIncludingExpired((ice, segment) -> {
-               if (ice.canExpire()) {
-                  // Have to synchronize on the entry to make sure we see the value and metadata at the same time
-                  boolean expiredMortal;
-                  boolean expiredTransient;
-                  V value;
-                  long lifespan;
-                  long maxIdle;
-                  synchronized (ice) {
-                     value = ice.getValue();
-                     lifespan = ice.getLifespan();
-                     maxIdle = ice.getMaxIdle();
-                     expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, ice.getCreated(), currentTimeMillis);
-                     expiredTransient = ExpiryHelper.isExpiredTransient(maxIdle, ice.getLastUsed(), currentTimeMillis);
+         LocalizedCacheTopology topology;
+         // Purge all contents until we know we did so with a stable topology
+         do {
+            topology = distributionManager.getCacheTopology();
+         } while (purgeInMemoryContents(topology));
+      }
+
+      if (!Thread.currentThread().isInterrupted()) {
+         persistenceManager.purgeExpired();
+      }
+   }
+
+   /**
+    * Purges in memory contents removing any expired entries.
+    * @return true if there was a topology change
+    */
+   private boolean purgeInMemoryContents(LocalizedCacheTopology topology) {
+      long start = 0;
+      int removedEntries = 0;
+      AtomicInteger errors = new AtomicInteger();
+      AtomicInteger outstandingExpirations = new AtomicInteger();
```

Comment:
I don't think you need to track outstanding expiration, the loop does a poll for each removed entry except the first MAX_CONCURRENT_EXPIRATIONS:

https://github.com/infinispan/infinispan/commit/3cbb8d0e4c57124ac632a870f63463593a955df8

Reply:
The problem is we need to be able to wait until all have finished. There is no way to poll for the last few to confirm it is completed otherwise.

Reply:
for example if we found 5 expirations, no caller would be able to be waiting on those, since we don't track those otherwise.

Reply:
https://github.com/infinispan/infinispan/commit/3cbb8d0e4c57124ac632a870f63463593a955df8#diff-a3098505167aeedeccbaf511fdf6f73fR167

Reply:
👍